clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -120,6 +123,7 @@ public Sender(KafkaClient client,
         this.clientId = clientId;
         this.sensors = new SenderMetrics(metrics);
         this.requestTimeout = requestTimeout;
+        this.metadataStaleMs = metadata.getMetadataMaxAge() + 3 * (requestTimeout + metadata.refreshBackoff());
```

Comment:
Added the following comment before the formula. I hope it makes sense. 

```
/* metadata becomes "stale" for batch expiry purpose when the time since the last successful update exceeds
 * the metadataStaleMs value. This value must be greater than the metadata.max.age and some delta to allow
 * a few retries. A small number of retries (3) are chosen because metadata retries have no upper bound. 
 * However, as retries are subject to both regular request timeout and the backoff, staleness determination
 * is delayed by that factor.
 */

```